### ISSUE TITLE
seeking about dataset

### ISSUE THREAD

**Issue by flwmxd** [buniatyann/MeshGNN#1]
your approach looks like using GNN score to replace QEM error when doing simplification ? right?

do you have any pre-trained dataset I can use for a test ? 
thanks.

**Comment by buniatyann** [comment]
Hello
Yes, the approach uses GNN-predicted scores as an alternative to QEM for edge collapse priority.
No, there is no pre-trained model or dataset included in this repository. You need to train the model yourself,  this library provides layers, optimizers, losses, training loop, but no ready-to-use weights. For training you can only use wavefront OBJ (.obj) files. How to train

#include <gnnmath/geometry/mesh.hpp>
#include <gnnmath/geometry/feature_extraction.hpp>
#include <gnnmath/gnn/pipeline.hpp>
#include <gnnmath/gnn/layers/gcn_layer.hpp>
#include <gnnmath/gnn/layers/edge_conv_layer.hpp>
#include <gnnmath/gnn/training.hpp>

using namespace gnnmath;

int main() {
    // 1. Load your mesh
    mesh::mesh m;
    m.load_obj("your_model.obj");

    // 2. Extract features (7D per node) and adjacency
    auto node_features = mesh::compute_combined_node_features(m);
    auto adj = m.to_adjacency_matrix();

    // 3. Build pipeline: 7 → 16 → 8 → 1
    auto pipeline = std::make_shared<gnn::pipeline>();
    pipeline->add_layer(std::make_unique<gnn::gcn_layer>(7, 16));
    pipeline->add_layer(std::make_unique<gnn::gcn_layer>(16, 8));
    pipeline->add_layer(std::make_unique<gnn::edge_conv_layer>(8, 1));

    // 4. Create trainer (Adam, lr=0.001, weight_decay=0.0001)
    gnn::trainer trainer(pipeline, 0.001, gnn::optimizer_type::ADAM, 0.0001);

    // 5. YOU must prepare your own target scores
    //    e.g. per-node scores derived from compute_quadric_error()
    std::vector<vector::vector> targets = /* your ground-truth */;

    // 6. Training loop
    for (int epoch = 0; epoch < 100; ++epoch) {
        trainer.train_step(node_features, adj, targets);
        auto predictions = pipeline->process(node_features, adj);
        double loss = trainer.compute_loss(predictions, targets);
        std::cout << "Epoch " << epoch << " Loss: " << loss << "\n";
    }

    // 7. Save your trained model
    pipeline->save("trained_model.bin");

    // 8. Later: load and use for simplification
    // pipeline->load("trained_model.bin");
    // auto scores = pipeline->process(node_features, adj);
    // mesh::simplify_gnn_edge_collapse(m, target_count, edge_scores);
}

The hardest part is defining your own training targets. One of the  practical approaches is to use compute_quadric_error() on each edge to get geometric importance scores, then derive per-node targets from those (because this trains the GNN to approximate the geometric heuristic but generalize across unseen meshes). 
Thank you

**Comment by buniatyann** [comment]
You should train on many meshes, not just one.
That's kind of the whole point of using a GNN here — if you only needed to simplify one specific mesh, you'd just run QEM directly and skip the neural network entirely.
The idea is that by training on a variety of shapes, the network learns general geometric intuition: "flat areas are cheap to collapse, sharp edges and high-curvature regions should be preserved." Then you can throw any new mesh at it and get decent results without recomputing everything from scratch.
So in practice you need to 1. Gather a diverse set of OBJ files (organic shapes, hard-surface models, architectural stuff, whatever your use case involves) 2. Generate ground-truth collapse scores for each (e.g. from QEM) 3. Train on all of them together 4. Save the model once, use it on anything
Here's what that looks like in code:

std::vector<std::string> training_meshes = {
    "data/mesh_01.obj",
    "data/mesh_02.obj",
    "data/mesh_03.obj",
    // the more variety the better
};

auto pipeline = std::make_shared<gnn::pipeline>();
pipeline->add_layer(std::make_unique<gnn::gcn_layer>(7, 16));
pipeline->add_layer(std::make_unique<gnn::gcn_layer>(16, 8));
pipeline->add_layer(std::make_unique<gnn::edge_conv_layer>(8, 1));

gnn::trainer trainer(pipeline, 0.001, gnn::optimizer_type::ADAM, 0.0001);

for (int epoch = 0; epoch < 100; ++epoch) {
    double total_loss = 0.0;

    for (const auto& path : training_meshes) {
        mesh::mesh m;
        m.load_obj(path);

        auto features = mesh::compute_combined_node_features(m);
        auto adj = m.to_adjacency_matrix();
        auto targets = /* your ground-truth scores for this mesh */;

        trainer.train_step(features, adj, targets);

        auto predictions = pipeline->process(features, adj);
        total_loss += trainer.compute_loss(predictions, targets);
    }

    std::cout << "Epoch " << epoch
              << " Avg Loss: " << total_loss / training_meshes.size() << "\n";
}

// one model to rule them all
pipeline->save("universal_model.bin");


If you train on just one mesh, the model basically memorizes it — which defeats the purpose. More variety in training = better generalization on meshes it's never seen before.

________________________________

**Comment by flwmxd** [comment]
it sounds interesting. I will take a look later. do you have any picture comparison ? Just curious about how it looks like.

**Comment by buniatyann** [comment]
> it sounds interesting. I will take a look later. do you have any picture comparison ? Just curious about how it looks like.

Unfortunately I don't have any visual comparisons to show — the library doesn't include any rendered before/after images.

But if you want to see the results yourself, it's pretty straightforward. Run the demo on the included Porsche model:

``` bash
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make -j$(nproc)
./meshgnn_lib data/Porsche_911_GT2.obj
```
This will output a simplified OBJ file. You can then open both the original and simplified meshes side by side in any 3D viewer — MeshLab, Blender, or even an online OBJ viewer — to see the difference yourself.
